analysis.ml

```diff
@@ -58,10 +46,111 @@ let forward_analysis (init_state : ('a * pc))
         | Some merged ->
             cell := Some merged;
             let updated = update pc merged in
-            let succs = successors program pc in
+            let succs = successors pc in
             let new_work = List.map (fun pc -> (updated, pc)) succs in
             work (new_work @ rest)
         end
   in
-  work [init_state];
+  work init_state;
   Array.map (!) program_state
+
+let exits program =
+  let rec exits pc : Pc.t list =
+    if Array.length program = pc then []
+    else
+      ((if List.length (successors program pc) = 0 then [pc] else [])
+        @ exits (pc + 1)) in
+  exits 0
+
+let forward_analysis_from init_pos init_state program =
+  let successors pc = successors program pc in
+  do_analysis successors [(init_state, init_pos)] program
+
+let forward_analysis init_state program =
+  forward_analysis_from 0 init_state program
+
+let backwards_analysis init_state program =
+  let exits = exits program in
+  let init_state = List.map (fun pc -> (init_state, pc)) exits in
+  let preds = predecessors program in
+  let predecessors pc = preds.(pc) in
+  do_analysis predecessors init_state program
+
+
+
+(* Use - Def style analysis
+ * Result as Map: variable -> set of instructions *)
+
+module InstrSet = Set.Make(Pc)
+module Def = struct
+  type t = variable
+  let compare = Variable.compare
+end
+module Defs = Map.Make(Def)
+
+
+let merge_defs =
+  let do_merge _ a b : InstrSet.t option =
+    match a, b with
+    | None, None -> None
+    | Some a, None -> Some a
+    | None, Some b -> Some b
+    | Some a, Some b -> Some (InstrSet.union a b) in
+  Defs.merge do_merge
+
+let equal_defs =
+  let do_equal a b = InstrSet.equal a b in
+  Defs.equal do_equal
+
+
+let reaching prog pc =
```

Comment:
I may be time to start adding documentation comments to say what those functions are doing. In OCaml it is customary to have the documentation comments just before the function declaration (without a blank line), and of the form `(** ... *)` (note the two starts at the beginning), with mere implementation comments of the form `(* ... *)`.